lib/net/x86-64-linux/Net.v3

```diff
@@ -150,59 +147,229 @@ def socketToSockAddr(socket: Socket, server: bool) -> Array<byte> {
 // messages.
 class Socket {
-	// IP address of socket
-	def addr: InetAddress;
 	// Associated file descriptor of socket
 	def var fd: int;
-	// Local port the socket is connected to
-	def var localPort: int;
+	// Is the socket an IPv6 socket?
+	def var isV6: bool;
+	// IP address of socket
+	def addr: InetAddress;
 	// Server port the socket is connected to
 	def var port: int;
-
-	new(addr, localPort, port, sockfd: int, createSocket: bool) {
+	// Local port the socket is connected to
+	def var localPort: int;
+	// Socket type
+	def socketType: SocketType;
+	// Has the socket been bound?
+	def var isBound = false;
+	// Is the socket listening for connections?
+	def var isListening = false;
+	// Is the sockect connected?
+	def var isConnected = false;
+
+	new(addr, port, localPort, socketType, sockfd: int, createSocket: bool) {
+		isV6 = InetAddress.Ipv6.?(addr);
 		if (createSocket) {
 			def family = if(InetAddress.Ipv4.?(addr), LinuxConst.AF_INET, LinuxConst.AF_INET6);
-			fd = Net.socket(family, LinuxConst.SOCK_STREAM /* type */, 0 /* protocol */);
+			def sType = if(socketType == SocketType.Tcp, LinuxConst.SOCK_STREAM, LinuxConst.SOCK_DGRAM);
+			fd = Net.socket(family, sType, 0 /* protocol */);
 		} else {
 			fd = sockfd;
 		}
 	}
 
+	def setBound(bound: bool) {
+		isBound = bound;
+	}
+
+	def setListening(listening: bool) {
+		isListening = listening;
+	}
+
+	def setConnected(connected: bool) {
+		isConnected = connected;
+	}
+
+	def getLocalAddress() -> InetAddress;
+
+	def getSockAddr() -> Array<byte>;
+
 	def setLocalPort(local: int) {
 		localPort = local;
 	}
 
-	// Read a message into buffer {buf} from the socket. Return how many bytes
-	// were read into the buffer.
-	def read(buf: Array<byte>) -> int {
-		return System.fileReadK(fd, buf, 0, buf.length);
+	def isUdp() -> bool {
+		return socketType == SocketType.Udp;
+	}
+
+	// Receive a message into buffer {buf} from the socket. Return how many bytes
+	// were read into the buffer. The receive will fail on a standard
+	// connectionless UDP socket.
+	def recv(buf: Array<byte>) -> int {
+		if (!isUdp() || isConnected) {
+			return System.fileReadK(fd, buf, 0, buf.length);
+		} else {
+			// socketType == SocketType.Udp
+			System.error("Error", "Called recv() on an unconnected UDP socket! Please use recvfrom() instead.");
+			return -1;
+		}
 	}
 
-	// Write a message {msg} into the socket.
-	def write(msg: string) {
-		System.fileWriteK(fd, msg, 0, msg.length);
+	// Receive a message into buffer {buf} from a remote host. Users can provide
+	// an instance of {packet} which will be overridden with information about
+	// the remote host. If no {packet} is provided, a new one is allocated and
+	// returned. Returns the number of bytes read into {buf} and information
+	// about the remote host for future communication. The receive will fail for
+	// a TCP socket.
+	def recvfrom(buf: Array<byte>, packet: UdpPacket) -> (int, UdpPacket) {
```

Comment:
If you accept `Range<byte>` here, users can pass sub-ranges of arrays (and `Array` auto-promotes to `Range`).

I'm thinking of going through most/all APIs that take arrays and upgrading them to ranges.

Reply:
Yeah I noticed this behaviour with the `rangeToInetAddress` function. It's very cool! Good idea to change it to a `Range` type